File: basilisk-wiki/sandbox/acastillo/output_fields/tests_profiles/test_profiles_slab.c
/**
# `profiles_slab.h` against `profiles.h`, where they must agree

[profiles_slab.h](../profiles/profiles_slab.h) bins each cell into a slab in
one traversal where [profiles.h](../profiles/profiles.h) point-samples every
slab in turn. The header claims the two coincide when the sampled region is at
one resolution and each slab is exactly one row of cells. This runs both on
the same field under exactly that geometry and compares the files they write,
column by column.

The field is the two-phase shear flow of [test_profiles.c](test_profiles.c):

  f = 1 for y < 0 (phase 1), f = 0 for y > 0 (phase 2)
  rho = rho1*f + rho2*(1-f), mu = mu1*f + mu2*(1-f)

  U_i(x,y) = alpha_shear*x + beta_i*y
  V_i(x,y) = -alpha_shear*y

with mu1*beta1 = mu2*beta2. There is no refinement here: the grid is uniform
even on `quadtree`/`octree`, which is the point -- the precondition is about
resolution, not about the grid type.

The sampling geometry is what makes the claim testable. With `del = L0/N`, the
slab centres are the cell centres, `hmin = Y0 + del/2`, `hmax = Y0 + L0 -
del/2` and `n = N`, so `deltahn` comes out to `del` and slab `j` holds cell row
`j` alone. The in-plane sample counts `m1` (and `m2`) are `N` over the same
inset range, so `foreach_region()` lands one sample on each cell centre.
Getting this wrong is the interesting failure: see
[test_profiles_slab_bias.c](test_profiles_slab_bias.c).

The two accumulate in different orders -- one over sampled planes, the other
over cells in grid order -- so `test_profiles_slab.py` compares them against a
tolerance rather than asserting bit-for-bit equality. In practice they come out
identical here, bar a column of the octree dissipation profile that is
analytically zero and so holds only roundoff. */

#define LEVEL 5

#include "navier-stokes/centered.h"
#include "two-phase.h"
#include "acastillo/output_fields/profiles/profiles.h"
#include "acastillo/output_fields/profiles/profiles_slab.h"

double tau = 1.;
double alpha_shear = 0.5;

int main(){

  L0 = 1.0;
  X0 = Y0 = Z0 = -L0 / 2;
  N = 1 << LEVEL;
  init_grid (N);

  rho1 = 3.; rho2 = 1.;
  mu1 = 3.; mu2 = 1.;
  double beta1 = tau / mu1;
  double beta2 = tau / mu2;

#if dimension == 3
  fraction (f, -z);
#else
  fraction (f, -y);
#endif

  foreach() {
    double beta = f[] ? beta1 : beta2;
    u.x[] = alpha_shear * x + beta * y;
    u.y[] = -alpha_shear * y;
#if dimension == 3
    u.z[] = alpha_shear * z + beta * x;
#endif
  }
  boundary ((scalar *){u});
  foreach()
    rhov[] = rho1 * f[] + rho2 * (1. - f[]);
  boundary ({rhov});

  /**
  One slab per cell row, one in-plane sample per cell centre. */
  double del = L0/N;
  #define SAMPLING \
    xmin = X0 + del/2., xmax = X0 + L0 - del/2., \
    ymin = Y0 + del/2., ymax = Y0 + L0 - del/2., \
    hmin = Z0 + del/2., hmax = Z0 + L0 - del/2., \
    n = N, m1 = N, m2 = N, mode = "w"

  scalar * list = {u.x, u.y, rhov};
  profile_foreach_region (list, unity, "region_scalar.asc", SAMPLING);
  profile_scalar_slab    (list, unity, "slab_scalar.asc",   SAMPLING);

  scalar * list1 = {u.x, u.y};
  scalar * list2 = {u.y, u.y};
  profile_product_foreach_region (list1, list2, unity, "region_product.asc",
                                  SAMPLING);
  profile_product_slab           (list1, list2, unity, "slab_product.asc",
                                  SAMPLING);

  profile_dissipation_foreach_region (u, unity, "region_dissipation.asc",
                                      SAMPLING);
  profile_dissipation_slab           (u, unity, "slab_dissipation.asc",
                                      SAMPLING);

  #undef SAMPLING

  /**
  The report goes to stderr, i.e. to the `log` diffed against
  `test_profiles_slab.ref`. */
  if (pid() == 0)
    system ("python3 ../test_profiles_slab.py "
            "--pair region_scalar.asc slab_scalar.asc "
            "--pair region_product.asc slab_product.asc "
            "--pair region_dissipation.asc slab_dissipation.asc 1>&2");
}

File: basilisk-wiki/sandbox/acastillo/output_fields/tests_profiles/test_profiles_slab.py
#!/usr/bin/env python3
"""
Checker for `profiles_slab.h`, shared by `test_profiles_slab.c` and
`test_profiles_slab_bias.c`. Each writes the same profile twice, once through
`profiles.h` and once through `profiles_slab.h`; this compares the two files
column by column.

--pair <a> <b>: the two must agree to roundoff. They accumulate in different
  orders -- one sums over sampled planes, the other over cells in grid order --
  so the comparison is against PAIR_TOL rather than bit-for-bit, even though on
  the present fixtures they mostly come out identical. Deviations are measured
  against the largest value in the file; see compare().

--differ <a> <b>: the same comparison where the header says the two are *not*
  interchangeable, because the sampled region spans more than one resolution.
  Passes when the deviation exceeds DIFFER_MIN, i.e. when the precondition is
  demonstrably load-bearing. If this ever fails, `profiles_slab.h` has become
  more general than its documentation claims and the header should say so.

Deviations are reported as an order of magnitude rather than a value: MPI
reductions sum in rank order, so the last digits move with the rank count and
a literal value would make the frozen reference depend on how many ranks the
test happened to run under.

Usage:
  python3 test_profiles_slab.py [--pair <a> <b>] [--differ <a> <b>] ...

Exits non-zero (and prints FAIL) if a --pair disagrees beyond PAIR_TOL or a
--differ agrees more closely than DIFFER_MIN.
"""
import math
import sys

PAIR_TOL = 1e-10
DIFFER_MIN = 1e-6


def read_profile(filename):
  """Read a profile file as a list of rows of floats, ignoring comments."""
  rows = []
  with open(filename) as f:
    for line in f:
      line = line.strip()
      if not line or line.startswith('#'):
        continue
      rows.append([float(v) for v in line.split()])
  return rows


def compare(file_a, file_b):
  """Max deviation between two profiles, relative to the file's overall scale.

  Returns (deviation, message); deviation is None if the files cannot be
  compared at all, in which case the message says why.
  """
  a, b = read_profile(file_a), read_profile(file_b)
  if not a or not b:
    return None, f'no data rows in {file_a if not a else file_b}'
  if len(a) != len(b):
    return None, f'row counts differ: {len(a)} vs {len(b)}'
  if len(a[0]) != len(b[0]):
    return None, f'column counts differ: {len(a[0])} vs {len(b[0])}'

  # Every deviation is measured against the largest value in the file, not
  # against the entry it came from. These profiles carry columns that are
  # analytically zero and hold nothing but roundoff; per-entry normalisation
  # makes such a column its own yardstick and turns 1e-17 against 1e-17 into
  # O(1). Against the file's own scale, a genuine error in even a small column
  # still lands far above PAIR_TOL.
  biggest = max(abs(v) for row in a for v in row)
  if biggest == 0.:
    biggest = 1.

  worst = 0.
  for row_a, row_b in zip(a, b):
    for va, vb in zip(row_a, row_b):
      dev = abs(va - vb)/biggest
      if dev > worst:
        worst = dev
  return worst, ''


def magnitude(x):
  """Order of magnitude of x, as a string, or '0' when x vanishes."""
  if x == 0.:
    return '0'
  return f'1e{round(math.log10(x))}'


def check_pair(file_a, file_b):
  """The two must agree. Returns 0 on success, 1 on failure."""
  dev, why = compare(file_a, file_b)
  if dev is None:
    print(f'FAIL: {file_a} vs {file_b}: {why}')
    return 1
  if dev > PAIR_TOL:
    print(f'FAIL: {file_a} vs {file_b} disagree, '
          f'max relative deviation ~{magnitude(dev)}')
    return 1
  print(f'PASS: {file_a} vs {file_b} agree to ~{magnitude(dev)}')
  return 0


def check_differ(file_a, file_b):
  """The two must disagree. Returns 0 on success, 1 on failure."""
  dev, why = compare(file_a, file_b)
  if dev is None:
    print(f'FAIL: {file_a} vs {file_b}: {why}')
    return 1
  if dev < DIFFER_MIN:
    print(f'FAIL: {file_a} vs {file_b} agree to ~{magnitude(dev)}; the '
          f'documented precondition looks unnecessary')
    return 1
  print(f'PASS: {file_a} vs {file_b} differ by ~{magnitude(dev)}, '
        f'as the precondition requires')
  return 0


CHECKS = {'--pair': check_pair, '--differ': check_differ}

USAGE = ('Usage: python3 test_profiles_slab.py [--pair <a> <b>] '
         '[--differ <a> <b>] ...')


def main():
  argv = sys.argv[1:]
  jobs = []  # (check_fn, file_a, file_b), in command-line order
  i = 0
  while i < len(argv):
    flag = argv[i]
    if flag not in CHECKS or i + 2 >= len(argv):
      print(USAGE)
      return 1
    jobs.append((CHECKS[flag], argv[i + 1], argv[i + 2]))
    i += 3

  if not jobs:
    print(USAGE)
    return 1

  status = 0
  for check, file_a, file_b in jobs:
    status |= check(file_a, file_b)
  return status


if __name__ == '__main__':
  sys.exit(main())
